Add support of scancode_cli to CoLic Backend ( WIP )

inishchith opened the pull request:
Addresses: #27 

@valeriocos In reference to your PR. This contains some minor changes. 
Still, there are some improvements which can be made and tests should be added accordingly.
Please do have a look, Thanks :)

Work attributions: @valeriocos 's #28

valeriocos commented on graal/backends/core/analyzers/scancode.py:
```diff
@@ -28,32 +28,39 @@
 from .analyzer import Analyzer
 
 
+SCANCODE_EXEC = "scancode-toolkit-3.0.0/scancode"
```
`scancode-toolkit-3.0.0` this could be removed ? if we keep it like this, the day another release is out we will have to modify the code, what do you think ?

inishchith replied:
Agreed. I'll make the changes

valeriocos commented on graal/backends/core/analyzers/scancode.py:
```diff
@@ -28,32 +28,39 @@
 
 
+SCANCODE_EXEC = "scancode-toolkit-3.0.0/scancode"
+SCANCODE_CLI_EXEC = "scancode-toolkit-3.0.0/etc/scripts/scancli.py"
```
same as above

valeriocos commented on graal/backends/core/colic.py:
```diff
@@ -101,13 +105,15 @@ def fetch(self, category=CATEGORY_COLIC_NOMOS, paths=None,
 
-        This backend generates two types of item which can be:
-        'code_license_nomos' or 'code_license_scancode'.
+        This backend generates three types of item which can be:
```
Would you think to change this comment with: `This backend generates the following types of item:`. In this way we don't have to remember of modifying this text if more categories are added.

inishchith replied:
Yes. Thanks for the suggestion.
I'll update this, Thanks

valeriocos commented:
Thank you for addressing the comments @inishchith

inishchith commented:
@valeriocos I've updated the PR. 
I've made the changes wrt. the assumption that `exec_path` provided would always point to the `scancode` executable. Please let me know if we're keeping it as the default convention of usage.

Thanks

valeriocos commented:
Sorry for the late reply @inishchith . 
To avoid confusion It is probably better if `exec_path` points to the file that should be executed (either `scancode` or scancli.py`). Why `exec_path` should always point to `scancode` ?

inishchith commented:
@valeriocos This came to my mind as I referred your implementation [here](https://github.com/valeriocos/grimoirelab-graal/commit/63a472fd54d6e491f55e731b11a4aec9c7b085e3#diff-88309ccaaf85d4fc1ec566c704a98b82R58) and thought if we were planning to do so, now i understand that it was for a quick test.
Thanks for the clarification. I'll make the changes

valeriocos commented:
perfect, thanks! That branch was a quick test, PR #28 should be better :)

inishchith commented:
@valeriocos I've made the changes. Please do have a look
I moving ahead to working on tests :)

valeriocos commented:
Great @inishchith , thank you for the updates. Please ping when the tests are ready! :)

inishchith commented:
@valeriocos I had a thought before starting work on tests.
 
We are required to accommodate configuring `scancli.py` before the tests start. So, I was thinking to perform [this]() checkout commit and `1st` and `2nd` step from [this doc](https://github.com/nexB/scancode-toolkit/commit/8afa686fb71b9540029234e5a40c0572c4457c28#diff-f826f8c8f6f35f368b2a692610f05d62R12) in build under `before_script` under travisci configuration.
Reason being, we're strictly using `scancode v3.0.0` for tests and as I tested it locally, both `Scancode` and `ScanCode cli` both work perfectly on the checkout commit.
Please let me know what you think :)

inishchith commented:
@valeriocos I'm not sure which there's no `travis-ci` and `coverage` triggered on last push.
Can you please check?

inishchith commented:
@valeriocos Please review when you get time
Thanks :)

valeriocos commented:
I'm on it @inishchith :)

valeriocos commented on graal/backends/core/analyzers/scancode.py:
```diff
@@ -62,9 +69,62 @@ def analyze(self, **kwargs):
+
+        :returns result: dict of the results of the analysis
+        """
 
```
There shouldn't be an empty line

valeriocos commented on graal/backends/core/analyzers/scancode.py:
```diff
@@ -62,9 +69,62 @@ def analyze(self, **kwargs):
+    def analyze(self, **kwargs):
+        """Add information about license
+
+        :param file_path: file path
```
This param doesn't seem to be clear enough (since it is valid only for scancode). Maybe we could add another param (`file_paths`) and add some text to explain when `file_path` and `file_paths` are valid. What do you think ?

inishchith replied:
Thanks for pointing that out. Yes, it's not clear as of now.
I was thinking something like:
```
:param file_path: file path ( in case of scancode )
:param file_paths: file paths ( in case of scancode_cli for concurrent execution on files )
```

What do you think?

valeriocos replied:
It looks OK! Let's go for this solution and if in the future we see it isn't clear we will change it (maybe scancode cli could be another analyzer).

inishchith replied:
Thanks for the quick response :)

valeriocos commented on graal/backends/core/analyzers/scancode.py:
```diff
@@ -22,38 +22,45 @@
-        """Add information about license
+        if self.cli:
+            exec_path = self.exec_path.replace(SCANCODE_CLI_EXEC, CONFIGURE_EXEC)
+            _ = subprocess.check_output([exec_path, '--help']).decode("utf-8")
```
it shouldn't be `_ = subprocess.check_output([exec_path])` (as reported here: https://github.com/nexB/scancode-toolkit/blob/develop/etc/scripts/scanserv.README#L14) ?

valeriocos replied:
I guess the param `--help` runs ./configure under the hood. Thank you for updating it.

valeriocos commented on graal/backends/core/colic.py:
```diff
@@ -101,13 +105,15 @@ def fetch(self, category=CATEGORY_COLIC_NOMOS, paths=None,
-        This backend generates two types of item which can be:
-        'code_license_nomos' or 'code_license_scancode'.
+        This backend generates the following types of item:
+        'code_license_nomos', 'code_license_scancode' or 'code_license_scancode_cli.
```
code_license_scancode_cli misses a `'` at the end. The line seems to long, maybe it could be rewritten as:
```
This backend generates the following types of item:
- 'code_license_nomos'
- 'code_license_scancode'
- 'code_license_scancode_cli'
```
What do you think ?

valeriocos commented on graal/backends/core/colic.py:
```diff
@@ -59,9 +61,9 @@ class CoLic(Graal):
+    version = '0.5.0'
 
-    CATEGORIES = [CATEGORY_COLIC_NOMOS, CATEGORY_COLIC_SCANCODE]
+    CATEGORIES = [CATEGORY_COLIC_NOMOS, CATEGORY_COLIC_SCANCODE, CATEGORY_COLIC_SCANCODE_CLI]
```
This line seems to long, maybe it could be rewritten as:
```
CATEGORIES = [CATEGORY_COLIC_NOMOS, 
              CATEGORY_COLIC_SCANCODE, 
              CATEGORY_COLIC_SCANCODE_CLI]
```
What do you think ?

valeriocos commented on tests/test_scancode.py:
```diff
@@ -43,18 +43,38 @@ def test_init(self):
 
         self.assertIn('licenses', result)
 
+    def test_analyze_scancode_cli(self):
```
The tests look perfect, but it's probably better to separate them in two different classe: TestScanCode and TestScanCodeCli, what do you think ?

inishchith replied:
@valeriocos Thanks for the suggestion.
I had a thought when I worked on this, I proceeded with the thought that they belong to the same analyzer. 
But as you mentioned above that we could (in future) separate both into different analyzers, It seems better to keep two different classes for the test.
I'll update this :)

inishchith commented:
@valeriocos 
I've gone through the comments and I see there are improvements needed.
I'll work on them and update you accordingly :)

Thanks

valeriocos commented:
Great! Thank you @inishchith to check the comments

inishchith commented:
@valeriocos I've worked on the requested changes. 

- [x] Merged (earlier) commits `1` & `3` into `1`, made the requested changes and updated commit description.
- [x] Updated the description of `2nd` commit (travis.yml)
- [x] Added installation and usage guide to `README`

Please let me know if I've missed something

Thanks

inishchith commented:
@valeriocos Please do have a look when you get time :)
Thanks

valeriocos commented on graal/backends/core/colic.py:
```diff
@@ -170,17 +190,20 @@ class LicenseAnalyzer:
 
     def analyze(self, file_path):
-        """Analyze the content of a file using Nomos
+        """Analyze the content of a file using Nomos/Scancode
```
@inishchith can you update the doc as done at: https://github.com/chaoss/grimoirelab-graal/pull/29/files#diff-88309ccaaf85d4fc1ec566c704a98b82R119 ?

inishchith commented:
@valeriocos Thanks for the review
I've updated the doc-string, please do have a look :)
